### ISSUE TITLE
Crash when server function throws redirect and wrapped in createAsync

### ISSUE THREAD

**Issue by apatrida** [solidjs/solid-router#429]
### Describe the bug

If you throw a redirect from a server function wrapped in createAsync it does not handle the redirect nor serialize it across.  Server functions called as actions might want to throw redirects, and if used elsewhere it is likely they would as well.  `cache` wrapping handles the redirect, but if you do not want caching, you do not have a `noCache` option and therefore throw or return a redirect causing what?  currently a crash.

Example:
```
async function serverApiCallThrow(isIndirect?: boolean): Promise<string> {
    'use server';
    throw redirect(`/${isIndirect ? 'fromIndirect' : 'fromThrow'}/whateva?q=456`);
    return 'ok'
}

...
  const asyncThing = createAsync(()=>serverApiCallThrow())
  
...
  <Show when={asyncThing()}>
      <div>boo!</div>
  </Show>
results in exception:

Error: Unknown error at castError (...server bundle...)

and client shows: 

<img width="804" alt="Screenshot 2024-05-26 at 8 26 41 AM" src="https://github.com/solidjs/solid-router/assets/182340/838cba0b-4002-4c14-bf55-83f3442a9c33">


### Your Example Website or App

see above

### Steps to Reproduce the Bug or Issue

1. create a server function
2. throw a redirect in that functino
3. use in a component with `createAsync`
4. boom

### Expected behavior

The redirect happens or thrown as error, but not a serialization crash

### Screenshots or Videos

_No response_

### Platform

any

### Additional context

_No response_

**Comment by Brendonovich** [comment]
From what I can tell this is already being thrown as a normal error and not a crash - wrapping the `<Show>` in an `ErrorBoundary` functions as you'd expect.
Perhaps just a better error message is necessary?

**Comment by maksimsemenov** [comment]
I think the issue here is that instead of handling the redirect, and navigating user to the new url, it throws an error. I don't think wrapping `<Show>` in `ErrorBoundary` will help.

[From the docs](https://docs.solidjs.com/solid-start/advanced/return-responses) it seems that only `cache` and `action` functions know how to correctly handle redirects. But I think there are cases, when we don't want to cache this function. For me it seems like missing functionality. We can try to use some workarounds, like cache revalidation, or pass some dummy random argument, so it will always produce unique cache key, but that doesn't seem as right thing to do.

**Comment by Brendonovich** [comment]
Yeah `cache` and `action` are the only wrappers that handle redirects, it's possible that a third neutral one is needed.
Even then a thrown response in a raw server function surfacing as an error is still expected behaviour, it could just be reported better.
Wrapping the `<Show>` in `ErrorBoundary` doesn't help, but it functions as you'd expect any other error to rather than causing a _crash_ which would be more worrying.

**Comment by sfnmk** [comment]
Just flagging here that I stumbled upon this problem as well. It'd be useful to have another more neutral way of sending redirects.

I was fetching some session data from the server and if it's not present I wanted to redirect to the sign in page. I ended up wrapping the function in an action but had to go through all the `useAction` / `useSubmission` overhead to be able to get the value. `createAsync` already returned a signal which is easier to handle.

**Comment by ryansolid** [comment]
Closing - the two halves of this have both resolved:

**The crash is stale.** The `castError: Unknown error` came from the old solid-start serialization path. The current `@solidjs/start` server-function client detects `Location`/`X-Revalidate`/`X-Single-Flight` responses and returns the raw `Response` to the caller instead of exploding during deserialization, and thrown errors come back as properly typed errors via `X-Error`.

**Raw server functions not auto-redirecting is by design.** A redirect is only actionable when the router is in the loop, and `query`/`action` are exactly those integration points - `query` for reads (it navigates on a thrown/returned redirect, applies `Set-Cookie`, etc.), `action` for writes. `createAsync` is deliberately router-agnostic: it's the bridge to the Solid 2.0 core primitive, which won't know about routing either, so it can never be the place redirects are handled.

For the "I don't want caching" concern in the thread: `query` is lighter than it sounds - it's request-scoped dedup on the server and a short preload/back-cache window on the client, not a persistent cache; `revalidate()`/`query.delete()` give you explicit control. Wrapping the read in `query` is the intended "neutral wrapper", and that's also where the thrown-redirect contract is guaranteed:

```ts
const getSession = query(async () => {
  "use server";
  const session = await fetchSession();
  if (!session) throw redirect("/signin");
  return session;
}, "session");

const session = createAsync(() => getSession());
```
